### PR TITLE
Enable more tt-train tests to run on Blackhole 

### PR DESCRIPTION
**Pull request by kevinwuTT**
### Problem description
In an example of a recent run of [Nightly tt-metal L2 tests workflow](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/), there are currently 4 nightly tt-train jobs and 1 non-nightly tt-train job.

#### Nightly jobs:
* [tt-train nightly tests wormhole_b0 N150](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/job/61599327216)
* [tt-train nightly tests wormhole_b0 N300](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/job/61599327473)
* [tt-train nightly tests blackhole P100](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/job/61599327245)
* [tt-train nightly tests blackhole P150b](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/job/61599327559)

In the batch of nightly jobs, various tests were skipped intentionally if running on BH. From my testing in current tt-metal, the following tests do pass when running on BH, so the restrictions will be lifted:
* CrossEntropyForwardTest.NIGHTLY_CrossEntropyForward_Huge_Forward
* CrossEntropyBackwardTest.NIGHTLY_CrossEntropyBackward_Huge_Backward
* RMSNormOpTest.NIGHTLY_RMSNorm_Forward_Batch
* RMSNormOpTest.NIGHTLY_RMSNorm_Backward_Batch
* RMSNormOpTest.NIGHTLY_RMSNorm_Compare_SmallBatch_NonUnit
* RMSNormOpTest.NIGHTLY_RMSNorm_Compare_Masking_Patterns
* SoftmaxTest.NIGHTLY_SoftmaxTest_Huge_Batch
* SDPAForwardTest.NIGHTLY_SDPAForwardTest_SmallBatch_12Heads_6Group
* SDPAForwardTest.NIGHTLY_SDPAForwardTest_Batch_12Heads_6Group

#### Non-nightly jobs:
* [tt-train wormhole_b0 N150](https://github.com/tenstorrent/tt-metal/actions/runs/21386468638/job/61599327317)

For non-nightly jobs, the workflow was configured to run tt-train tests on Wormhole N150 only, regardless of whether Blackhole was selected. This PR will enable non-nightly jobs to also run on N300, P100a, and P150b. 

In addition, 2 intentionally disabled tests appear to run fine on current tt-metal, so the restriction will also be lifted for these tests. However, 1 test seem to show consistent accuracy issues on BH, but passes on WH. So that test will be skipped for BH.

Enabled:
* SerializationConfigs/FlatBufferFileSerializationTest.ScopedTempDirWriteReadRoundTrip/FLOAT32_ROW_MAJOR_DEVICE
* SerializationConfigs/FlatBufferFileSerializationTest.ScopedTempDirWriteReadRoundTrip/FLOAT32_ROW_MAJOR_HOST

Skipped:
* TrivialTnnFixedTest.TestSamplingPositiveTemperatureWithMask
* SoftmaxBackward/SoftmaxBackwardOpTypedTest.SoftmaxBackward_SubCoreGrid_Rectangular/bf16

#### Other changes:
Comments are added for the following tests. They were disabled due to ongoing bugs, but should be enabled once they're fixed.
* NanoLlamaTest.NIGHTLY_Default
* DropoutTest.TestKeepRatioApproximatelyNormal

### Checklist

- [x] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-kevinwu/tt-train-tests-comments)
  - Latest with Watcher enabled: https://github.com/tenstorrent/tt-metal/actions/runs/21918079985
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ctr-kevinwu/tt-train-tests-comments)
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-kevinwu/tt-train-tests-comments)

### CODE REVIEW

File: .github/workflows/tt-metal-l2-nightly.yaml [hunk 1]
```diff
@@ -220,6 +221,18 @@ jobs:
...
           echo "matrix-tutorials=$matrix_tutorials" >> $GITHUB_OUTPUT
 
+          # Matrix for tt-train-cpp-unit-tests (N150, N300, P100a, P150b)
+          matrix_tt_train='[]'
+          if [[ "$run_wormhole" == "true" ]]; then
+            matrix_tt_train=$(add_entry "$matrix_tt_train" "wormhole_b0" "N150")
+            matrix_tt_train=$(add_entry "$matrix_tt_train" "wormhole_b0" "N300")
+          fi
+          if [[ "$run_blackhole" == "true" ]]; then
+            matrix_tt_train=$(add_entry "$matrix_tt_train" "blackhole" "P100a")
```
Comment (kevinwuTT) [inline .github/workflows/tt-metal-l2-nightly.yaml:231]: If I set this to "P100" instead of "P100a", docker seems to fail consistently when pulling images. Not really sure why since other matrix variables have "P100".

https://github.com/tenstorrent/tt-metal/actions/runs/21527259106/job/62048781363
```
  /usr/bin/docker pull harbor.ci.tenstorrent.net/ghcr.io/tenstorrent/tt-metal/tt-metalium/ubuntu-22.04-dev-amd64:a07d7ede36967a6f9478ac68926621ce34ecc0d4
  Error response from daemon: Get "https://harbor.ci.tenstorrent.net/v2/": dial tcp: lookup harbor.ci.tenstorrent.net on 127.0.0.53:53: server misbehaving
```

https://github.com/tenstorrent/tt-metal/actions/runs/21527259106/job/62034604205
```
  /usr/bin/docker pull harbor.ci.tenstorrent.net/ghcr.io/tenstorrent/tt-metal/tt-metalium/ubuntu-22.04-dev-amd64:a07d7ede36967a6f9478ac68926621ce34ecc0d4
  Error response from daemon: Get "https://harbor.ci.tenstorrent.net/v2/": dial tcp: lookup harbor.ci.tenstorrent.net on 127.0.0.53:53: read udp 127.0.0.1:44061->127.0.0.53:53: i/o timeout
```

File: tt-train/tests/model/nano_gpt_test.cpp [hunk 2]
```diff
@@ -24,6 +24,7 @@ namespace {
 Nightly tests could be enabled by setting the environment variable ENABLE_NIGHTLY_TT_TRAIN_TESTS=1
 or setting 'is_nigthly_tt_train_tests_enabled' variable to true.
 */
+// TODO: Disabled by https://github.com/tenstorrent/tt-metal/pull/26877. Revert once fix is implemented.
```
Comment (jmalone-tt) [inline tt-train/tests/model/nano_gpt_test.cpp]: Can we add a tracking issue for implementing the fix and mention that here instead of the PR that disabled these tests?

File: tt-train/tests/ttnn_fixed/dropout_op_test.cpp [hunk 3]
```diff
@@ -86,6 +86,7 @@ TEST_F(DropoutTest, TestProb) {
 }
 
 TEST_F(DropoutTest, TestKeepRatioApproximatelyNormal) {
+    // TODO: Enable once RNG is improved.
```
Comment (jmalone-tt) [inline tt-train/tests/ttnn_fixed/dropout_op_test.cpp]: I think @athompsonTT worked on a SIMD RNG - not sure if that would allow us to enable this test

File: tt-train/tests/ttnn_fixed/trivial_ttnn_ops_test.cpp [hunk 4]
```diff
@@ -275,6 +276,11 @@ TEST_F(TrivialTnnFixedTest, TestSamplingPositiveTemperatureNoMask) {
 }
 
 TEST_F(TrivialTnnFixedTest, TestSamplingPositiveTemperatureWithMask) {
+    // TODO: Accuracy issue with BH. Remove once issue is fixed.
```
Comment (jmalone-tt) [inline tt-train/tests/ttnn_fixed/trivial_ttnn_ops_test.cpp]: Let's make a tracking issue for this and add to the backlog

**Review by jmalone-tt — COMMENTED** [review]
Thanks for working on this. I'm curious to see how the recent changes to disable tests when watcher is enabled will impact things. Ideally, we would still be able to run all tests, we might need to change whatever enables watcher though

File: tt-train/tests/ttnn_fixed/dropout_op_test.cpp (hunk 3 above)
Comment (kevinwuTT) [reply tt-train/tests/ttnn_fixed/dropout_op_test.cpp]: I saw some PRs that were closed

https://github.com/tenstorrent/tt-metal/pull/33105
https://github.com/tenstorrent/tt-metal/pull/32050

If I have the time, I can try these locally and see if they affect the test. I posted a tracking issue here: https://github.com/tenstorrent/tt-metal/issues/38149

**Review by zbaczewskiTT — APPROVED** [review]

File: .github/workflows/tt-metal-l2-nightly.yaml (hunk 1 above)
Comment (williamlyTT) [reply .github/workflows/tt-metal-l2-nightly.yaml:231]: @kevinwuTT the correct runner to target is `P100a`.
P100 will target the wrong set of runners for this workflow, you can delete this line: https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/tt-train-post-commit.yaml#L61

File: tt-train/tests/ops/rmsnorm_op_test.cpp [hunk 5]
```diff
@@ -91,6 +90,9 @@ TEST_F(RMSNormOpTest, RMSNorm_Small_Backward) {
 }
 
 TEST_F(RMSNormOpTest, NIGHTLY_RMSNorm_Forward_Batch) {
+    // Skip with watcher enabled github issue #37193
```
Comment (mdragulaTT) [inline tt-train/tests/ops/rmsnorm_op_test.cpp]: I fixed that issue. Are you sure those tests are still need to be skipped with watcher?

File: tt-train/tests/ttnn_fixed/dropout_op_test.cpp [hunk 6]
```diff
@@ -86,6 +86,7 @@ TEST_F(DropoutTest, TestProb) {
 }
 
 TEST_F(DropoutTest, TestKeepRatioApproximatelyNormal) {
+    // TODO: Enable once RNG is improved. Tracking issue: https://github.com/tenstorrent/tt-metal/issues/38149
```
Comment (mdragulaTT) [inline tt-train/tests/ttnn_fixed/dropout_op_test.cpp:89]: Isn't related to @athompsonTT work on SIMD RNG? I didn't fully follow the status of that work. Can anyone help if Adam fixed that before he left or we have a still open issue about it? @jmalone-tt @ayerofieiev-tt @philei-tt

File: tt-train/tests/model/nano_gpt_test.cpp [hunk 7]
```diff
@@ -24,6 +24,7 @@ namespace {
 Nightly tests could be enabled by setting the environment variable ENABLE_NIGHTLY_TT_TRAIN_TESTS=1
 or setting 'is_nigthly_tt_train_tests_enabled' variable to true.
 */
+// TODO: Disabled due to instability. Tracking issue: https://github.com/tenstorrent/tt-metal/issues/37337
```
Comment (mdragulaTT) [inline tt-train/tests/model/nano_gpt_test.cpp]: This is vague. Same as the issue description. Instability of what actually? Those comments are added but it is not clear what is disabled actually

**Review by mdragulaTT — COMMENTED** [review]
I left a few comments—hope they’re helpful. Please address them or let me know if you disagree, and I’ll approve the PR

File: tt-train/tests/ops/rmsnorm_op_test.cpp (hunk 5 above)
Comment (kevinwuTT) [reply tt-train/tests/ops/rmsnorm_op_test.cpp]: I removed the skips and ran RMSNormOpTest with 

```
export TT_METAL_WATCHER=2
export TT_METAL_WATCHER_APPEND=1
export TT_METAL_WATCHER_NOINLINE=1
./build/tt-train/tests/ttml_tests --gtest_filter=RMSNormOpTest*
```

And they all pass on a WH LB. Will mark this as resolved. Thanks.

File: tt-train/tests/ttnn_fixed/dropout_op_test.cpp (hunk 6 above)
Comment (jmalone-tt) [reply tt-train/tests/ttnn_fixed/dropout_op_test.cpp:89]: Looks like the last merge he had regarding SIMD_RNG was to disable it by default:
https://github.com/tenstorrent/tt-metal/pull/35144

I think the new issue to track this is good

**Review by jmalone-tt — APPROVED** [review]
Change looks good to me as long as regressions are passing in CI. Agree with Maciek's note about clarifying reason in nano gpt test

**Review by mdragulaTT — APPROVED** [review]

**Comment by blozano-tt** [comment]
@copilot Can you please open a PR to address the earlier PR feedback?

> No timeout passed from l2-nightly
The tt-train-cpp-unit-tests job calls tt-train-post-commit.yaml but does not pass a timeout: input. The reusable workflow uses timeout-minutes: ${{ inputs.timeout }}, which evaluates to empty → GitHub defaults to 6h. A hung BH shard will hold a scarce P100a/P150b runner for up to 6 hours. Suggest adding timeout: 90 (or similar) to the with: block in the l2-nightly caller.

**Review by blozano-tt — APPROVED** [review]
Approved to unblock. I've asked copilot to propose an adjustment to the timeout scheme. Please consider that feedback.
